Adding DiscreteData Module

AnkeAnke opened the pull request:
**State**
- The basic setup is done, there is at least one example for abstract classes
- Has worked with one project (KTHInviwo, Percolation)
- Tests should be extended
- Lots of places to discuss, please comment!
-- Iterators & cache
-- Connection iteration data
-- DataSet with const Channels only?

**What environment has the code been compiled and tested on?** 
- Windows
- VC 2017

AnkeAnke committed:
Initial commit
Adding module, empty CMakeList

AnkeAnke committed:
Initial Channel, DataChannel<T>

AnkeAnke committed:
Adding channel types, iterator
* BufferChannel (data explicit in std::vector)
* AnalyticChannel (data implicit by std::function)
* ChannelIterator (holds implicit data if needed)

AnkeAnke committed:
Adding namespace 'dd', formatting

AnkeAnke committed:
Adding Connectivity, Iterators
* Connectivity (containing incidence from all dimensions to all others)
** connections currently via std::vector, find better variant
* ElementIterator (go through all elements in one GridDimension)
* ConnectionIterator (go through all neighbors of an element)

AnkeAnke committed:
Adding StructuredGrid, util

AnkeAnke committed:
Fixing forgotten license headers

AnkeAnke committed:
Forgotten variable refactors

AnkeAnke committed:
Adding DataSet

AnkeAnke committed:
Restructuring files into folders

AnkeAnke committed:
Fixing includes after restructure

AnkeAnke committed:
Fixed rename errors

AnkeAnke committed:
Adding tests
Rather sketchy but good to get templates compiled

AnkeAnke committed:
Copyright year

petersteneteg commented on modules/discretedata/channels/analyticchannel.h:
```diff
@@ -0,0 +1,122 @@
+
+    // Methods
+public:
+    ind getNumElements() const override { return numElements_; }
```
We usually use size for the number of elements, to be consistant with stl etc. makes writing generic code easier...

AnkeAnke replied:
I'd rather stay with this.
We decided to have ind (signed long long) everywhere to keep consistent in our code. Negative values as invalid marker, and the agreement that signed is the way to go.
Now, we could go with size_t in this one case, but that is not obvious to the outside user and then there is a mixed interface again.

petersteneteg replied:
ind is fine with me, I was more referring to the name of the function

AnkeAnke replied:
Misunderstood you there.
As we have two different sizes (components and elements) I like it better this explicit. If there was only one I'd fully agree.

petersteneteg replied:
But then if you would let T represent the whole type not just the element type, there would only be one size :)

AnkeAnke replied:
It's size now.

petersteneteg commented on modules/discretedata/channels/analyticchannel.h:
```diff
@@ -0,0 +1,122 @@
+class AnalyticChannel : public DataChannel<T> {
+    // Types
+public:
+    typedef typename std::function<void(T*, ind)> Function;
```
Having the dimension of vec undefined does not seem very typesafe

AnkeAnke replied:
I'll have to think and look through this again, implement numComponents_ as a template arg and find out whether it works. I know we had a reason against it, but I'm not sure what it was and whether it is still valid ;)

petersteneteg replied:
One way would also be to just let T be what ever type you want, Float, double, glm::vec3, std::array<float, 17>, myGreatTensorType, or what ever, instead of just float or double like now. That way you don’t have to add the number of components as a template argument at all it would rather be built into T.

AnkeAnke replied:
For the current setup that would work. But then having more channel types (like unsteay, hich just adds a 1.0 at the end, or another compound type) would be more tedious again.

petersteneteg replied:
Not sure I agree here I believe that leaving all the detail up to the type T would give you the most freedom in this case. doing something like adding a 1.0 would be trivial in that case. Maybe you can expand more on what the problem would be?

AnkeAnke replied:
Scenario: I'm integrating inertial particles. You have a 3D position, 3D velocity, and a 1D 1.0 at the end. I'd like to have a
`UnsteadyChannel<CompoundChannel<BufferChannel<float>, BufferChannel<float>>>`. All have type float, and the overall size would be 7. Representing this with Vec3 in there could be cumbersome.

petersteneteg replied:
Interesting, I think I understand what you want to achieve. But what would happen if for example the positions where in doubles and the velocity in floats? how would this api handle that?

I would to something like
```
CompoundChannel<BufferChannel<dvec3>, BufferChannel<vec3>, AnalyticalChannel<float>>
```
and instad of returning setting a pointer to a float. I would return a std::tuple<dvec3, vec3, float> or maybe std::tuple<dvec3&,  vec3&, float> if the data should be modifiable

AnkeAnke replied:
Your tuple could not be processed like a math vector though. How does one normalize a vector of float and int? If different types of data are needed, they can be stored in different channels. 

The advantage that remains is the ability to store Tensors without the need to reinterpret a large vector.

We'll be discussing whether numComponents should be a template argument on Monday.

AnkeAnke replied:
GLM does not have a template-sized vector or matrix type as far as I see. So I'll include some other vectors.

petersteneteg replied:
It does not at the our version, but in the latest master the dimension is a template param. We should update at some point. In the mean time we have used a template alias 
https://github.com/inviwo/inviwo/blob/3764b5095d98707c8897eac6b16645df9419af5e/include/inviwo/core/util/glm.h#L743-L747
Our does not work beyond N = 4 though...

petersteneteg commented on modules/discretedata/channels/analyticchannel.h:
```diff
@@ -0,0 +1,122 @@
+*  Exported Types                                               *
+*--------------------------------------------------------------*/
+
+typedef AnalyticChannel<float>      AnalyticChannelFloat;
```
We prefer the more readable using x = y

petersteneteg commented on modules/discretedata/channels/bufferchannel.h:
```diff
@@ -0,0 +1,199 @@
+    *   @param name Name associated with the channel
+    *   @param definedOn GridPrimitive the data is defined on, default: 0D vertices
+    */
+    BufferChannel(std::vector<T>& data, ind numComponents, const std::string& name,
```
const vector& ?

AnkeAnke replied:
Aye. 12bd8db10050f840bac722f260cab0376ef31a25

petersteneteg commented on modules/discretedata/channels/bufferchannel.h:
```diff
@@ -0,0 +1,199 @@
+
+    // Methods
+public:
+    virtual ind getNumElements() const override;
```
size?

petersteneteg commented on modules/discretedata/channels/bufferchannel.h:
```diff
@@ -0,0 +1,199 @@
+    *   @param index Linear point index
+    *   @return Pointer to data, size T[NumComponents]
+    */
+    const T* operator[](ind index) const { return get(index); }
```
I don't like these raw pointers to array without a explict size.

petersteneteg commented on modules/discretedata/channels/bufferchannel.h:
```diff
@@ -0,0 +1,199 @@
+/*--------------------------------------------------------------*
+*  Exported Types                                               *
+*--------------------------------------------------------------*/
+
```
Using

petersteneteg commented on modules/discretedata/channels/bufferchannel.h:
```diff
@@ -0,0 +1,199 @@
+void BufferChannel<T>::fill(T* const dest, const ind index) const {
+    ivwAssert(index >= 0 && index < getNumElements(), "Index out of bounds: " << index);
+
+    memcpy(dest, buffer_.data() + (index * numComponents_), sizeof(T) * numComponents_);
```
This limmits T to trivial types, should maybe be noted, or just use std::copy, will always work and will be optimized to memcpy for trivial types

petersteneteg commented on modules/discretedata/channels/channeliterator.h:
```diff
@@ -0,0 +1,107 @@
+    // Members
+protected:
+    /** Index to the current element */
+    ind Index;
```
Lowercase members

petersteneteg commented on modules/discretedata/channels/channeliterator.h:
```diff
@@ -0,0 +1,107 @@
+    ~ChannelIterator() { delete[] Data; }
+
+    /** Dereference to get data */
+    T* operator*();
```
How do I get the number of components?

petersteneteg replied:
Should return a vec<N>& I think using some vec type, glm, std::arrray, span, or some other

petersteneteg replied:
We should probably discuss if you shoukd use glm or some oher vec type soon, it will probably influence the design alot

petersteneteg commented on modules/discretedata/channels/channeliterator.h:
```diff
@@ -0,0 +1,107 @@
+    //*** Bidirectional Iteration ***\\
+
+    /** Walk forward */
+    ChannelIterator& operator++() {
```
Post increment?

petersteneteg commented on modules/discretedata/channels/channeliterator.h:
```diff
@@ -0,0 +1,107 @@
+template <typename T>
+class ChannelIterator {
+public:
+    ChannelIterator(DataChannel<T>* parent, ind index)
```
should define iteractor category tag types. see http://en.cppreference.com/w/cpp/iterator/iterator_traits

petersteneteg commented on modules/discretedata/channels/channeliterator.inl:
```diff
@@ -0,0 +1,55 @@
+    ivwAssert(Parent, "No channel to iterate is set.");
+
+    // If our parent is a buffer, return a direct data pointer.
+    if (dynamic_cast<BufferChannel<T>*>(Parent))
```
it's probably a big perf hit to do a dynamic cast on every iterator dereference. This functionality probably has to be in the channel not in the iterator.

petersteneteg commented on modules/discretedata/channels/channeliterator.inl:
```diff
@@ -0,0 +1,55 @@
+
+    // No buffer, evaluate analytics.
+    // Is the data up to date?
+    if (DataIndex != Index) {
```
So in this case the channel is analytical and hence const? right, so there is really no reason to provide a non const iterator since it can't be modified anyway. And if it cant be modified you could just return the "T" by value not having to go through all this trouble of have a local buffer holding the value so you can point to it.

We will probably need to provide a ConstChannelIterator and a ChannelIterator. Not even sure if the mutable one is needed since all the data is usually immutable anyway? at least the const one seems the more important one.

petersteneteg commented on modules/discretedata/channels/datachannel.cpp:
```diff
@@ -0,0 +1,78 @@
+
+Channel::Channel(ind numComponents, const std::string& name, DataFormatId format,
+                 GridPrimitive prim)
+    : numComponents_(numComponents) {
```
In my thinking we only use the meta data for "extra" information not usually in the data structure,  I would rather store name, format, primitive as regular class members.

petersteneteg commented on modules/discretedata/channels/datachannel.cpp:
```diff
@@ -0,0 +1,78 @@
+
+    // Create and set format identifier.
+    IntMetaData* formatData = createMetaData<IntMetaData>("DataFormatId");
+    *formatData = (int)format;
```
One should never use c-style cast in c++

petersteneteg commented on modules/discretedata/channels/datachannel.h:
```diff
@@ -0,0 +1,171 @@
+*   Assign channels to any dimensions this way.
+*   If these do not suffice, cast the respective short.
+*/
+enum GridPrimitive : char {
```
use enum classs

petersteneteg commented on modules/discretedata/channels/datachannel.h:
```diff
@@ -0,0 +1,171 @@
+namespace dd {
+
+/** Discretedata index type **/
+typedef signed long long ind;
```
use using

petersteneteg commented on modules/discretedata/channels/datachannel.h:
```diff
@@ -0,0 +1,171 @@
+    // Attributes
+public:
+    virtual ind getNumElements() const = 0;
+    ind getNumComponents() const { return numComponents_; }
```
the dataformat also has the number of components...

petersteneteg commented on modules/discretedata/channels/datachannel.h:
```diff
@@ -0,0 +1,171 @@
+
+    // Attributes
+public:
+    virtual ind getNumElements() const = 0;
```
use size

petersteneteg commented on modules/discretedata/channels/datachannel.h:
```diff
@@ -0,0 +1,171 @@
+    *   @param dest Position to write to, expect T[NumComponents]
+    *   @param index Linear point index
+    */
+    virtual void fill(T* dest, ind index) const = 0;
```
Imho this kind of interface with T* is nice very safe, should really be avoided.

petersteneteg commented on modules/discretedata/channels/datachannel.h:
```diff
@@ -0,0 +1,171 @@
+    *   @param dest Position to write to, expect T[NumComponents]
+    *   @param index Linear point index
+    */
+    void operator()(T* dest, ind index) const { fill(dest, index); }
```
Same as above

petersteneteg commented on modules/discretedata/channels/datachannel.h:
```diff
@@ -0,0 +1,171 @@
+
+    ChannelIterator<T> begin();
+
+    ChannelIterator<T> end();
```
should also have const iterators

petersteneteg commented on modules/discretedata/channels/datachannel.h:
```diff
@@ -0,0 +1,171 @@
+
+    /** Returns the "DataFormatId" meta data */
+    DataFormatId getDataFormatId() const;
+
```
you might want to add support for dispatching on the dataformat here similar to VolumeRAM for example

petersteneteg commented on modules/discretedata/channels/datachannel.inl:
```diff
@@ -0,0 +1,72 @@
+DataChannel<T>::DataChannel(ind numComponents, const std::string& name, GridPrimitive definedOn)
+    : Channel(numComponents, name, DataFormatId::NotSpecialized, definedOn) {
+    // Switch all types.
+    if (std::is_same<T, f16>::value) setDataFormatId(DataFormatId::Float16);
```
you can just use ```DataFormat<T>::id()```

petersteneteg commented on modules/discretedata/connectivity/connectioniterator.h:
```diff
@@ -0,0 +1,154 @@
+class IVW_MODULE_DISCRETEDATA_API ConnectionIterator {
+    friend ConnectionIterator operator+(ind, ConnectionIterator&);
+    friend ConnectionIterator operator-(ind, ConnectionIterator&);
+
```
missing iterator tag types

petersteneteg commented on modules/discretedata/connectivity/connectioniterator.h:
```diff
@@ -0,0 +1,154 @@
+    //*** Bidirectional Iteration ***\\
+
+    /** Walk forward */
+    ConnectionIterator& operator++() {
```
missing post increment

ResearchDaniel commented on modules/discretedata/connectivity/connectivity.h:
```diff
@@ -0,0 +1,128 @@
+    *   @param dim Dimension of element (edge, face, volume...)
+    *   @Param index Index of respective element type
+    */
+    virtual double getPrimitiveMeasure(GridPrimitive, ind) const { return -1; }
```
should be abstract?
ind?

AnkeAnke replied:
Is not abstract so far as it is nice to have a volume computation given, but not absolutely necessary (hence, standard implementation with invalid output).

ind is a typedef for signed long long, which is used for all indices, iteration variables etc consistently in the project. We decided on signed (faster?, possibility to return 'invalid value' as in this case, consistent type should prevent unnecessary casting). The typedef is in datachannel.h.

ResearchDaniel commented on modules/discretedata/connectivity/structuredgrid.cpp:
```diff
@@ -0,0 +1,264 @@
+    if (from == to && from == gridDimension_) {
+        // Linear Index to nD Cell Index.
+        ind idxCutoff = idxLin;
+        std::vector<ind> index(numCellsPerDimension_.size(), -1);
```
Any reason to initialize to -1?
reserve might be faster...

AnkeAnke replied:
This way we have he possibility to check invalid values. I made use of it in the loose implementation of unstructured data, for example (not uploaded yet). Also, this iterates over dimensions. The highest I've ever needed is 7, so I expect <100 dimensions, paired with it happening only once each new grid is initialized, I would prefer comfort over speed here.

ResearchDaniel replied:
The loop afterwards, `index[dim] = idxCutoff % dimSize;`, overwrites all values anyway?
No biggie though

AnkeAnke replied:
Ooooh. Right. Well, yes.

ResearchDaniel commented on modules/discretedata/connectivity/structuredgrid.cpp:
```diff
@@ -0,0 +1,264 @@
+            break;
+
+        default:
+            LogWarn("Data type not supported. Edit " << __FILE__);
```
Probably better to throw exception

AnkeAnke replied:
Good point.
On another hand I am still wondering whether we should try to keep as much as possible of the module `noexcept` for performance, maybe even explicitly. Though I'm not sure how far that can go with stl bits everywhere. So far I've relied on asserts mostly.
Either way, this code will change to use Peter's dispatch functionality soon.

ResearchDaniel replied:
I would try and see if it has any performance impact in that case. This case might bite you in the *** later :D

AnkeAnke replied:
Yes, probably a very premature opt. thought. I'll add in dispatch and see how that handles it.

petersteneteg replied:
I would not worry about the cost of exceptions here at all...

ResearchDaniel commented on modules/discretedata/connectivity/structuredgrid.h:
```diff
@@ -0,0 +1,110 @@
+
+    @author Anke Friederici and Tino Weinkauf
+*/
+class IVW_MODULE_DISCRETEDATA_API StructuredGrid : public Connectivity {
```
SpatialData?
https://github.com/inviwo/inviwo/blob/69cca9b6ecd58037bda0ed9e6f53d02f189f19a7/include/inviwo/core/datastructures/spatialdata.h

AnkeAnke replied:
We had a look into it, but mainly two reasons against it:
- We find the model, world transforms more confusing than anything else. Though I guess it might be beneficial to have them, but did not yet find a scenario where I need them.
- It is possible to have a 1D structured grid (a line) in a 3D space. I think that can not be represented by SpatialData?

I'm very open to integrate them if there is a benefit. I don't know where they are used so far.
Any thoughts?

AnkeAnke replied:
Also, the grid so far is independent of the point dimension. So this is probably more relevant for the DataSet - might actually be a good place to save the transformations.

ResearchDaniel replied:
Was also thinking that it might be more suitable in the DataSet. 
The SpatialData is used to place the data in world space. I guess the vertices must be given in world space right now?

AnkeAnke replied:
Well, there is no way to display it so far... ;) Currently there is only one space.
We would not scale positions etc. into a unit cube or similar but keep it in the original space. So I'd say we have the data in model space, deriving from spatial entity would add a world space. 

When two data sets are in different scales/units, the relative scaling should probably be done via the model space matrix. But we'd then have to be careful about scaling the data (e.g., velocity should always be in the same space as the positions, while temperature would not be scaled). Difficult topic.

petersteneteg replied:
The spatial model could be implements as a spatial position channel if needed.

ResearchDaniel reviewed (commented):
Looks good! Some questions and comments

petersteneteg commented on modules/discretedata/connectivity/connectioniterator.h:
```diff
@@ -0,0 +1,154 @@
+    ConnectionRange(ind fromIndex, GridPrimitive fromDim, GridPrimitive toDim,
+                    const Connectivity* parent);
+
+    ConnectionIterator begin() { return ConnectionIterator(parent_, toDimension_, connections_, 0); }
```
Const?

petersteneteg commented on modules/discretedata/connectivity/connectivity.h:
```diff
@@ -0,0 +1,128 @@
+    // Methods
+public:
+    /** \brief Returns the maximal dimension of the grid */
+    GridPrimitive getDimension() const { return gridDimension_; }
```
I find dimension to be a bit confusing here, it that the best name? would Rank, or Dimensionality be better? Or is this the standard name in the field? I think dimension is often confused with NumElements?

petersteneteg commented on modules/discretedata/connectivity/connectivity.h:
```diff
@@ -0,0 +1,128 @@
+    *   @param elementType Type to get number of
+    *   @return Number of elements
+    */
+    virtual ind getNumElements(GridPrimitive elementType) const;
```
would prefer size

petersteneteg commented on modules/discretedata/connectivity/connectivity.h:
```diff
@@ -0,0 +1,128 @@
+    *   @param to Dimension the result lives in
+    *   @return All connected indices in dimension to
+    */
+    virtual std::vector<ind> getConnections(ind index, GridPrimitive from,
```
are there any limitation of the combination of to and from, or is all permutations valid?

petersteneteg commented on modules/discretedata/connectivity/elementiterator.h:
```diff
@@ -0,0 +1,143 @@
+class IVW_MODULE_DISCRETEDATA_API ElementIterator {
+    friend ElementIterator operator+(ind, ElementIterator&);
+    friend ElementIterator operator-(ind, ElementIterator&);
+
```
iterator tag types missing

petersteneteg commented on modules/discretedata/connectivity/elementiterator.h:
```diff
@@ -0,0 +1,143 @@
+    //*** Bidirectional Iteration ***\\
+
+    /** Walk forward */
+    ElementIterator& operator++() {
```
post increment missing

petersteneteg commented on modules/discretedata/connectivity/elementiterator.h:
```diff
@@ -0,0 +1,143 @@
+    ~ElementIterator() = default;
+
+    /** Dereference to 'get data' */
+    ElementIterator operator*() const;
```
return index here?

petersteneteg commented on modules/discretedata/connectivity/elementiterator.h:
```diff
@@ -0,0 +1,143 @@
+    const Connectivity* parent_;
+
+    /** GridPrimitive type iterated over (0D vertices etc) */
+    const GridPrimitive dimension_;
```
Question it the GridPrimitive the same as the Polytope (https://en.wikipedia.org/wiki/Polytope)? I think it would be nice if we could find a different naming scheme here then using dimension, since it's meaning is a bit ambiguous

petersteneteg replied:
or is a Simplex type? https://en.wikipedia.org/wiki/Simplex there are so many names for these things :)

petersteneteg commented on modules/discretedata/connectivity/connectivity.h:
```diff
@@ -0,0 +1,128 @@
+
+public:
+    /** Vertex positions */
+    std::shared_ptr<const Channel> vertices_;
```
Does this have to be in the base class? should the derived classes not define this them self?

petersteneteg commented on modules/discretedata/connectivity/structuredgrid.cpp:
```diff
@@ -0,0 +1,264 @@
+    : Connectivity(gridDimension), numCellsPerDimension_(numCellsPerDim) {
+    ivwAssert(numCellsPerDim.size() == gridDimension,
+              "Grid dimension should match cell dimension.");
+    numCellsPerDimension_ = std::vector<ind>(numCellsPerDim);
```
again?

petersteneteg commented on modules/discretedata/connectivity/structuredgrid.h:
```diff
@@ -0,0 +1,110 @@
+template <typename T>
+double StructuredGrid::computeHexVolume(ind index) const {
+    // Work with respective type
+    std::shared_ptr<const DataChannel<T>> doubleVertices =
```
not used?

petersteneteg commented on modules/discretedata/connectivity/structuredgrid.cpp:
```diff
@@ -0,0 +1,264 @@
+
+    numGridPrimitives_[gridDimension] = numCells;
+    numGridPrimitives_[GridPrimitive::Vertex] = numVerts;
+}
```
Don't use need to initialize the Connectivity::vertices_ member? or where is that done?

petersteneteg commented on modules/discretedata/connectivity/structuredgrid.cpp:
```diff
@@ -0,0 +1,264 @@
+    double measure = -1;
+
+    // TODO: Make this kind of code obsolete or at least compact.
+    switch (vertices_->getDataFormatId()) {
```
here you could use the dispatch thing

petersteneteg commented on modules/discretedata/dataset.h:
```diff
@@ -0,0 +1,209 @@
+namespace inviwo {
+namespace dd {
+
+typedef std::shared_ptr<Channel> SharedChannel;
```
From my experience, i think it's better not to typedef the shared_ptr like this, the user really needs to know that it is a std::shared_ptr to be able to create is using for example std::make_shared. And also for how to think about it with respect to lifetime etc. so I would prefer to use std::shared_ptr<Channel> std::shared_ptr<const Channel> in the interface instead of the typedefs.

petersteneteg commented on modules/discretedata/dataset.h:
```diff
@@ -0,0 +1,209 @@
+    }
+};
+// Map used for storing and querying channels by name and GridPrimitive type.
+typedef std::map<
```
define inside of DataChannelMap?

petersteneteg commented on modules/discretedata/dataset.h:
```diff
@@ -0,0 +1,209 @@
+    *   @param channel Pointer to data, takes memory ownership
+    *   @return Shared pointer for further handling
+    */
+    SharedChannel addChannel(Channel* channel);
```
In an new think like this, I don't think we need to supply the raw pointer interface. the user should always use std::make_shared to create the channels anyway. So I think this overload can be removed.

AnkeAnke committed:
Tiny optimization

petersteneteg commented on modules/discretedata/dataset.h:
```diff
@@ -0,0 +1,209 @@
+
+    /** Returns the first channel from an unordered list.
+    */
+    SharedConstChannel getFirstChannel() const;
```
What's the use case for getting the first one? is that a special one?

petersteneteg commented on modules/discretedata/dataset.h:
```diff
@@ -0,0 +1,209 @@
+    /** Number of channels currently held
+    */
+    ind numChannels() const { return ChannelSet.size(); }
+
```
can I ask for all the existing names?

AnkeAnke replied:
Currently the map is public anyways. But a list of available channels would be a better solution, I agree.

AnkeAnke replied:
Added in e9b5b392941ed6bef3a2c9ed9c938f68114dae9a

AnkeAnke replied:
Added in e9b5b392941ed6bef3a2c9ed9c938f68114dae9a

petersteneteg commented on modules/discretedata/dataset.h:
```diff
@@ -0,0 +1,209 @@
+    */
+    template <typename T>
+    std::shared_ptr<const DataChannel<T>> getChannel(
+        const char* name, GridPrimitive definedOn = GridPrimitive::Vertex) const;
```
don't think we need overloads for const char*

AnkeAnke replied:
I like to be able to just throw in a "name" without explicitly generating a std::string.

You talked about these basic_string_views, I'll look into it.

petersteneteg replied:
Don't think you need to worry about that performance impact. Would say that is premature optimisation. But you still find it important, then from the api point of view you should probably have that overload for all of the functions that takes string arguments, like the non template version and getAsBuffer...
About std::string_view since it is c++17 we don't use it in core so far. since we only require c++14. But since this is a completely new module and not a mandatory one for running inviwo... I think we could require c++17 in here if you want to. Would be kind of a waste not using the lastest standard when doing this from scratch...

AnkeAnke replied:
I remember there was some reason. Compiling it with actually working code right now.

AnkeAnke replied:
Actually, the error was the missing `const`. Works fine now. Removed.

petersteneteg commented on modules/discretedata/dataset.h:
```diff
@@ -0,0 +1,209 @@
+    /// Returns a const typed shared pointer to the grid, if casting is possible.
+    template <typename T>
+    std::shared_ptr<const T> getGrid() const {
+        return std::dynamic_pointer_cast<const T, const Connectivity>(Grid);
```
you don't need the second template arg to dynamic_pointer_cast it is deduced

AnkeAnke replied:
Removed. 414bc9f9dd1e522163234430364c04fae58d4a48

petersteneteg commented on modules/discretedata/dataset.h:
```diff
@@ -0,0 +1,209 @@
+    virtual ~DataSet() = default;
+
+    /** Default copy shares Channels and Connectivity */
+    DataSet(const DataSet& copy) = default;
```
what about assignment and move?

AnkeAnke replied:
All should be default. Would you prefer to have all 5 there explicitly or none?

petersteneteg replied:
loop up the guideline regarding the "role of five" or the "rule of zero"

petersteneteg commented on modules/discretedata/util.h:
```diff
@@ -0,0 +1,69 @@
+/** See std::hash_combine implementation from Boost
+*/
+template <class T>
+inline void CombineHash(std::size_t& hash, const T& value) {
```
think we have this in inviwo-core also...

AnkeAnke replied:
Any clue where I should look...?

petersteneteg replied:
sure
https://github.com/inviwo/inviwo/blob/a385ae059c75338576940b2a2f866a75a6dbc865/include/inviwo/core/util/stdextensions.h#L669-L677

petersteneteg reviewed (changes requested):
Really nice work!! there are many minor issues, and some interface things that I think we might need to discuss. But a very nice start I think!

petersteneteg commented:
Btw you should probably add the DiscreteData module to the jenkinsfile (about here https://github.com/inviwo/inviwo/blob/b068ac59c57658a38ec4c27ed10bc9a08c0065d1/Jenkinsfile#L80-L99) so it gets compiled in jenkins.

petersteneteg committed:
Enable DiscreteData in Jenkins

AnkeAnke committed:
Changing typedefs to usings

AnkeAnke committed:
Adding const

AnkeAnke committed:
DataChannelMap: not exposing raw map anymore
* getChannelNames return keys to all existing channels
* getChannel now overloaded for these keys (pairs of name and dimension)
* Minor fixes

AnkeAnke committed:
Fixing Linux errors/warnings
* No 2 phase lookup in Windows, should heighten test coverage of templates
* -Wpedantic does not like extra semicola

AnkeAnke committed:
Removing unnecessary functions
* pair hash combine
* const char* overload

AnkeAnke committed:
Remove unnecessary cast

petersteneteg commented:
I have now updated to the latest GLM verison....

AnkeAnke committed:
Adding template component size, template getting
* All data channels now have template <T,N>
* fill, get, iterators are templated with their return type
** Added layer of ChannelGetters inside of iterators to cope with this

AnkeAnke committed:
Small fixes

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/inviwo/inviwo?pullRequest=108) <br/>All committers have signed the CLA.

AnkeAnke committed:
Fixing iteration and getting, adding const iterator

AnkeAnke committed:
Fixing template errors
Compiling locally with MSVC hides a lot of errors that two-phase lookup brings up

AnkeAnke committed:
Fixing templates

AnkeAnke committed:
Replaced i64 by Tino

AnkeAnke committed:
Fix

AnkeAnke committed:
Fixed missing iterator const

AnkeAnke committed:
Valid partial specialization of DataChannel
I discovered sfinae. Old implementation was invalid, as template specialization is only valid by dependent type, not by value-dependent expressions (i.e., N = 1).

AnkeAnke committed:
Valid partial BufferChannel specialization

AnkeAnke committed:
Fix

AnkeAnke committed:
Fixing/simplifying buffer access by friending BufferGetter and BufferChannel

AnkeAnke committed:
Small fixes

AnkeAnke committed:
Removing core dependency: removing ivw_assert
Should probably have own assert at one point, c-assert is quite ugly

AnkeAnke committed:
Fix

AnkeAnke committed:
StructuredGrid restructure & fix

AnkeAnke committed:
Fixes for non-MSVC
MSVC compiles a class with some methods discarded completely by substitution failures, GCC does not.

AnkeAnke committed:
Template fixes

AnkeAnke committed:
Template fixes

AnkeAnke committed:
Restructuring again
Getting standard-template arguments into DataChannel and BufferChannel

AnkeAnke committed:
Small additions and formatting
* Switching template order in AnalyticChannel
* Not returning std::vector anymore, but filling provided one

AnkeAnke committed:
Periodic grid
* Adding structured grid with wrapping dimensions

AnkeAnke committed:
Test

AnkeAnke committed:
Adding example code as unit test

AnkeAnke committed:
"Fix", have to fix this test.

AnkeAnke committed:
Changing GridPrimitive to enum class
* And small fixes

AnkeAnke committed:
Changing namespace dd to discretedata

AnkeAnke committed:
Moving cell size computation outside
* Connectivity now knows the geometric cell type
** Cell type copied from vtk so far
* euclidean namespace contains tools to compute sizes of such

AnkeAnke committed:
Removing vertices_ from connectivity

AnkeAnke committed:
Removing indirection in DataSet

AnkeAnke committed:
Requiring a grid in DataSet creation
* Construction via any Connectivity or parameters for a StructuredGrid

AnkeAnke committed:
Removing/formatting comments

AnkeAnke committed:
Hiding Channel set. Fixing Jenkins warning, fixing template bugs

AnkeAnke committed:
Fixing template code

AnkeAnke committed:
Removing remnants of old code, template fix

AnkeAnke committed:
Fix

petersteneteg commented:
Great work!, Looking forward to start playing with it!
